README.md
# Introduction

This is one of a series of Python packages for
the Zope-based knowledge management platform 
[loops](http://www.cyberconcepts.de) (in German).

It contains functionalities for registration and 
authentication of members of a loops site. 

There will be new modules as well as modules migrated from 
the current loops and cybertools repositories.

For more information see
[www.cyberconcepts.org](https://www.cyberconcepts.org).

More to come...


CHANGES.txt
Changelog
=========

1.0.1 - 2017-01-02
------------------

- First public release


src/cco/member/tests.py
#! /usr/bin/python

"""
Tests for the 'cco.member' package.
"""

import os
import unittest, doctest
from zope import component
from zope.app.testing.setup import placefulSetUp, placefulTearDown
from zope.publisher.browser import TestRequest

from loops.interfaces import IConceptManager
from loops.tests.setup import TestSite


def setUp(self):
    site = placefulSetUp(True)
    t = TestSite(site)
    concepts, resources, views = t.setup()
    loopsRoot = site['loops']
    self.globs['loopsRoot'] = loopsRoot


def tearDown(self):
    placefulTearDown()


class Test(unittest.TestCase):
    "Basic tests."

    def testBasicStuff(self):
        pass


def test_suite():
    flags = doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS
    return unittest.TestSuite((
        unittest.makeSuite(Test),
        doctest.DocFileSuite('README.txt', optionflags=flags,
                     setUp=setUp, tearDown=tearDown),
        ))

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


src/cco/member/auth.py
#
#  Copyright (c) 2015 Helmut Merz helmutm@cy55.de
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

"""
Specialized authentication components.
"""

import hashlib
import logging
import random
from datetime import datetime, timedelta
from email.MIMEText import MIMEText
from urllib import urlencode
from zope.app.component import hooks
from zope.interface import Interface, implements
from zope import component
from zope.pluggableauth.interfaces import IAuthenticatedPrincipalFactory
from zope.pluggableauth.plugins.session import SessionCredentialsPlugin \
                            as BaseSessionCredentialsPlugin
from zope.pluggableauth.plugins.session import SessionCredentials
from zope.publisher.interfaces.http import IHTTPRequest
from zope.session.interfaces import ISession
from zope.traversing.browser import absoluteURL
from zope.traversing.namespace import view

from loops.browser.node import getViewConfiguration
from loops.organize.interfaces import IPresence
from loops.organize.party import getAuthenticationUtility
from loops.util import _


TIMEOUT = timedelta(minutes=5)
#PRIVKEY = "6LcGPQ4TAAAAABCyA_BCAKPkD6wW--IhUicbAZ11"   # for captcha
log = logging.getLogger('cco.member.auth')


class AuthURLNameSpace(view):

    def traverse(self, name, ignored):
        self.request.shiftNameToApplication()
        # ignore, has already been evaluated by credentials plugin
        return self.context


class TwoFactorSessionCredentials(SessionCredentials):

    def __init__(self, login, password):
        self.login = login
        self.password = password
        self.tan = random.randint(100000, 999999)
        self.timestamp = datetime.now()
        rng = range(len(str(self.tan)))
        t1 = random.choice(rng)
        rng.remove(t1)
        t2 = random.choice(rng)
        self.tanA, self.tanB = sorted((t1, t2))
        self.hash = (hashlib.
                        sha224("%s:%s:%s" % (login, password, self.tan)).
                        hexdigest())
        self.validated = False


class SessionCredentialsPlugin(BaseSessionCredentialsPlugin):

    tan_a_field = 'tan_a'
    tan_b_field = 'tan_b'
    hash_field = 'hash'

    def extractCredentials(self, request):
        if not IHTTPRequest.providedBy(request):
            return None
        login = request.get(self.loginfield, None)
        password = request.get(self.passwordfield, None)
        session = ISession(request)
        sessionData = session.get('zope.pluggableauth.browserplugins')
        traversalStack = request.getTraversalStack()
        authMethod = 'standard'
        credentials = None
        if sessionData:
            credentials = sessionData.get('credentials')
            if isinstance(sessionData, TwoFactorSessionCredentials):
                authMethod = '2factor'
        if (authMethod == 'standard' and 
                traversalStack and traversalStack[-1].startswith('++auth++')):
            authMethod = traversalStack[-1][8:]
        viewAnnotations = request.annotations.setdefault('loops.view', {})
        viewAnnotations['auth_method'] = authMethod
        #log.info('authentication method: %s.' % authMethod)
        if authMethod == 'standard':
            return self.extractStandardCredentials(
                            request, login, password, session, credentials)
        elif authMethod == '2factor':
            return self.extract2FactorCredentials(
                            request, login, password, session, credentials)
        else:
            return None

    def extractStandardCredentials(self, request, login, password, 
                                   session, credentials):
        if login and password:
            credentials = SessionCredentials(login, password)
        if credentials:
            sessionData = session['zope.pluggableauth.browserplugins']
            if credentials != sessionData.get('credentials'):
                sessionData['credentials'] = credentials
        else:
            return None
        return {'login': credentials.getLogin(),
                'password': credentials.getPassword()}

    def extract2FactorCredentials(self, request, login, password, 
                                  session, credentials):
        tan_a = request.get(self.tan_a_field, None)
        tan_b = request.get(self.tan_b_field, None)
        hash = request.get(self.hash_field, None)
        if (login and password) and not (tan_a or tan_b or hash):
            return self.processPhase1(request, session, login, password)
        if (tan_a and tan_b and hash) and not (login or password):
            credentials = self.processPhase2(request, session, hash, tan_a, tan_b)
        if credentials and credentials.validated:
            return {'login': credentials.getLogin(),
                    'password': credentials.getPassword()}
        return None

    def processPhase1(self, request, session, login, password):
        sessionData = session['zope.pluggableauth.browserplugins']
        credentials = TwoFactorSessionCredentials(login, password)
        sessionData['credentials'] = credentials
        # send email
        log.info("Processing phase 1, TAN: %s. " % credentials.tan)
        params = dict(h=credentials.hash, 
                      a=credentials.tanA+1, b=credentials.tanB+1)
        url = self.getUrl(request, '2fa_tan_form.html', params)
        return request.response.redirect(url, trusted=True)

    def processPhase2(self, request, session, hash, tan_a, tan_b):
        def _validate_tans(a, b, creds):
            tan = str(creds.tan)
            return tan[creds.tanA] == a and tan[creds.tanB] == b
        sessionData = session['zope.pluggableauth.browserplugins']
        credentials = sessionData.get('credentials')
        if not credentials:
            msg = 'Missing credentials'
            return log.warn(msg)
        log.info("Processing phase 2, TAN: %s. " % credentials.tan)
        if credentials.hash != hash:
            msg = 'Illegal hash.'
            return log.warn(msg)
        if credentials.timestamp < datetime.now() - TIMEOUT:
            msg = 'Timeout exceeded.'
            request.form['loops.message'] = msg
            return log.warn(msg)
        if not _validate_tans(tan_a, tan_b, credentials):
            msg = 'TAN digits not correct.'
            log.warn(msg)
            params = dict(h=credentials.hash, 
                          a=credentials.tanA+1, b=credentials.tanB+1)
            params['loops.message'] = msg
            url = self.getUrl(request, '2fa_tan_form.html', params)
            request.response.redirect(url)
            return None
        credentials.validated = True
        log.info('Credentials valid.')
        sessionData['credentials'] = credentials
        if request.get('camefrom'):
            request.response.redirect(request['camefrom'])
        return credentials

    def getUrl(self, request, action, params):
        if request.get('camefrom'):
            params['camefrom'] = request['camefrom']
        baseUrl = request.get('base_url') or ''
        if baseUrl and not baseUrl.endswith('/'):
            baseUrl += '/'
        return '%s%s?%s' % (baseUrl, action, urlencode(params))

    def challenge(self, request):
        if not IHTTPRequest.providedBy(request):
            return False
        site = hooks.getSite()
        path = request['PATH_INFO'].split('/++/')[-1] # strip virtual host stuff
        if not path.startswith('/'):
            path = '/' + path
        camefrom = request.getApplicationURL() + path
        if 'login' in camefrom:
            camefrom = '/'.join(camefrom.split('/')[:-1])
        url = '%s/@@%s?%s' % (absoluteURL(site, request),
                              self.loginpagename,
                              urlencode({'camefrom': camefrom}))
        request.response.redirect(url)
        return True

    def logout(self, request):
        presence = component.getUtility(IPresence)
        presence.removePresentUser(request.principal.id)
        super(SessionCredentialsPlugin, self).logout(request)


def getCredentials(request):
    session = ISession(request)
    sessionData = session.get('zope.pluggableauth.browserplugins')
    if not sessionData:
        return None
    return sessionData.get('credentials')


def getPrincipalFromCredentials(context, request, credentials):
    if not credentials:
        return None
    cred = dict(login=credentials.getLogin(), 
                password=credentials.getPassword())
    auth = getAuthenticationUtility(context)
    authenticatorPlugins = [p for n, p in auth.getAuthenticatorPlugins()]
    for authplugin in authenticatorPlugins:
        if authplugin is None:
            continue
        info = authplugin.authenticateCredentials(cred)
        if info is None:
            continue
        info.authenticatorPlugin = authplugin
        principal = component.getMultiAdapter((info, request),
            IAuthenticatedPrincipalFactory)(auth)
        principal.id = auth.prefix + info.id
        return principal



src/cco/member/pwpolicy.py
#
#  Copyright (c) 2016 Helmut Merz helmutm@cy55.de
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

"""
Check passwords for conformance to password policy.
"""


def checkPassword(pw):
    if len(pw) < 8:
        return False
    safety = dict(upper=False, lower=False, nonalpha=False)
    for c in pw:
        if c.isupper():
            safety['upper'] = True
        elif c.islower():
            safety['lower'] = True
        else:
            safety['nonalpha'] = True
    return False not in safety.values()


src/cco/member/interfaces.py
#
#  Copyright (c) 2016 Helmut Merz helmutm@cy55.de
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

"""
Interfaces for member registration, password change, etc.
"""

from zope.i18nmessageid import MessageFactory
from zope.interface import Interface
from zope import schema


_ = MessageFactory('cco.member')


class IPasswordChange(Interface):

    oldPassword = schema.Password(title=_(u'label_old_password'),
                    description=_(u'desc_old_password'),
                    required=True,)

    password = schema.Password(title=_(u'label_new_password'),
                    description=_(u'desc_new_password'),
                    required=True,)

    passwordConfirm = schema.Password(title=_(u'label_confirm_new_password'),
                    description=_(u'desc_confirm_new_password'),
                    required=True,)

    oldPassword.nostore = True
    password.nostore = True
    passwordConfirm.nostore = True



src/cco/member/browser.py
#
#  Copyright (c) 2016 Helmut Merz helmutm@cy55.de
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

"""
Login, logout, unauthorized stuff.
"""

from email.MIMEText import MIMEText
from zope.app.exception.browser.unauthorized import Unauthorized as DefaultUnauth
from zope.app.pagetemplate import ViewPageTemplateFile
from zope.app.security.interfaces import IAuthentication
from zope.app.security.interfaces import ILogout, IUnauthenticatedPrincipal
from zope.cachedescriptors.property import Lazy
from zope import component
from zope.i18n import translate
from zope.i18nmessageid import MessageFactory
from zope.interface import implements
from zope.publisher.interfaces.http import IHTTPRequest
from zope.sendmail.interfaces import IMailDelivery

from cco.member.auth import getCredentials, getPrincipalFromCredentials
from cco.member.interfaces import IPasswordChange
from cco.member.pwpolicy import checkPassword
from cybertools.composer.schema.browser.common import schema_macros
from cybertools.composer.schema.browser.form import Form
from cybertools.composer.schema.schema import FormState, FormError
from loops.browser.concept import ConceptView
from loops.browser.node import NodeView, getViewConfiguration
from loops.common import adapted
from loops.organize.interfaces import IMemberRegistrationManager
from loops.organize.party import getPersonForUser


_ = MessageFactory('cco.member')

template = ViewPageTemplateFile('auth.pt')


class LoginConcept(ConceptView):

    @Lazy
    def macro(self):
        return template.macros['login_form']


class LoginForm(NodeView):

    @Lazy
    def macro(self):
        return template.macros['login_form']

    @Lazy
    def item(self):
        return self

    @Lazy
    def isVisible(self):
        return self.isAnonymous

    def update(self, topLevel=True):
        if 'SUBMIT' in self.request.form and not self.isAnonymous:
            self.request.response.redirect(self.topMenu.url)
            return False
        return True


class TanForm(LoginForm):

    @Lazy
    def macro(self):
        return template.macros['tan_form']

    @Lazy
    def credentials(self):
        return getCredentials(self.request)

    def sendTanEmail(self):
        if self.credentials is None:
            return None
        person = None
        principal = getPrincipalFromCredentials(
                            self.context, self.request, self.credentials)
        if principal is not None:
            person = adapted(getPersonForUser(
                            self.context, self.request, principal))
        if person is None:     # invalid credentials
            # TODO: display message
            return None
        tan = self.credentials.tan
        recipient = person.email
        recipients = [recipient]
        lang = self.languageInfo.language
        subject = translate(_(u'tan_mail_subject'), target_language=lang)
        message = translate(_(u'tan_mail_text_$tan', mapping=dict(tan=tan)),
                            target_language=lang)
        senderInfo = self.globalOptions('email.sender')
        sender = senderInfo and senderInfo[0] or 'info@loops.cy55.de'
        sender = sender.encode('UTF-8')
        msg = MIMEText(message.encode('UTF-8'), 'plain', 'UTF-8')
        msg['Subject'] = subject.encode('UTF-8')
        msg['From'] = sender
        msg['To'] = ', '.join(recipients)
        mailhost = component.getUtility(IMailDelivery, 'Mail')
        mailhost.send(sender, recipients, msg.as_string())
        return recipient


class Logout(object):

    implements(ILogout)

    def __init__(self, context, request):
        self.context = context
        self.request = request

    def __call__(self):
        nextUrl = self.request.get('nextURL') or self.request.URL[-1]
        if not IUnauthenticatedPrincipal.providedBy(self.request.principal):
            auth = component.getUtility(IAuthentication)
            ILogout(auth).logout(self.request)
        return self.request.response.redirect(nextUrl)


class LogoutView(NodeView):

    @Lazy
    def body(self):
        nextUrl = self.topMenu.url
        if not IUnauthenticatedPrincipal.providedBy(self.request.principal):
            auth = component.getUtility(IAuthentication)
            ILogout(auth).logout(self.request)
        return self.request.response.redirect(nextUrl)


class Unauthorized(ConceptView):

    isTopLevel = True

    def __init__(self, context, request):
        self.context = context
        self.request = request

    def __call__(self):
        response = self.request.response
        response.setStatus(403)
        # make sure that squid does not keep the response in the cache
        response.setHeader('Expires', 'Mon, 26 Jul 1997 05:00:00 GMT')
        response.setHeader('Cache-Control', 'no-store, no-cache, must-revalidate')
        response.setHeader('Pragma', 'no-cache')
        if self.nodeView is None:
            v = DefaultUnauth(self.context, self.request)
            return v()
        url = self.nodeView.topMenu.url
        if self.isAnonymous:
            response.redirect(url)
        else:
            response.redirect(url + '/unauthorized')


class PasswordChange(NodeView, Form):

    interface = IPasswordChange
    message = _(u'message_password_changed')

    formErrors = dict(
        confirm_nomatch=FormError(_(u'error_password_confirm_nomatch')),
        wrong_oldpw=FormError(_(u'error_password_wrong_oldpw')),
        invalid_pw=FormError(_(u'error_password_invalid_pw')),
    )

    label = label_submit = _(u'label_change_password')

    @Lazy
    def macro(self):
        return schema_macros.macros['form']

    @Lazy
    def item(self):
        return self

    @Lazy
    def data(self):
        return dict(oldPassword=u'', password=u'', passwordConfirm=u'')

    def update(self):
        form = self.request.form
        if not form.get('action'):
            return True
        formState = self.formState = self.validate(form)
        if formState.severity > 0:
            return True
        pw = form.get('password')
        if not checkPassword(pw):
            fi = formState.fieldInstances['password']
            fi.setError('invalid_pw', self.formErrors)
            formState.severity = max(formState.severity, fi.severity)
            return True
        pwConfirm = form.get('passwordConfirm')
        if pw != pwConfirm:
            fi = formState.fieldInstances['password']
            fi.setError('confirm_nomatch', self.formErrors)
            formState.severity = max(formState.severity, fi.severity)
            return True
        oldPw = form.get('oldPassword')
        regMan = IMemberRegistrationManager(self.loopsRoot)
        principal = self.request.principal
        result = regMan.changePassword(principal, oldPw, pw)
        if not result:
            fi = formState.fieldInstances['oldPassword']
            fi.setError('wrong_oldpw', self.formErrors)
            formState.severity = max(formState.severity, fi.severity)
            return True
        url = '%s?loops.message=%s' % (self.url, self.message)
        self.request.response.redirect(url)
        return False

    def validate(self, data):
        formState = FormState()
        for f in self.schema.fields:
            fi = f.getFieldInstance()
            value = data.get(f.name)
            fi.validate(value, data)
            formState.fieldInstances.append(fi)
            formState.severity = max(formState.severity, fi.severity)
        return formState

